### PR TITLE
Add transactions search

### PR DESCRIPTION
**Pull request by AnnaGolosova**
## Link to ClickUp ticket

https://app.clickup.com/t/1jvea9b

## Describe this PR

Search endpoint needs to be moved to the Housing Search API - https://github.com/LBHackney-IT/financial-transactions-api/blob/development/FinancialTransactionsApi/V1/Gateways/ElasticSearch/SearchGateway.cs
The implementation of this was not agreed/discussed prior to implementation

### *Follow up actions after merging PR*

Transactions elasticSearch index should be created. We can use `Add transactions index` request for this purposes. This request is stored in HIT postman collection:
image

### CODE REVIEW

File: HousingSearchApi/V1/Boundary/Responses/Transactions/GetTransactionListResponse.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+using System.Collections.Generic;
+using Hackney.Shared.HousingSearch.Domain.Transactions;
+
+namespace HousingSearchApi.V1.Boundary.Responses.Transactions
+{
+    public class GetTransactionListResponse
+    {
+        private long _total;
+
+        public List<Transaction> Transactions { get; set; }
```
Comment (amidomichael) [inline HousingSearchApi/V1/Boundary/Responses/Transactions/GetTransactionListResponse.cs:10]: I would make this private and use the static factory method approach, and manipulate the list (by adding more items) through a method within this class.

File: HousingSearchApi/V1/Interfaces/SearchGateway.cs [hunk 2]
```diff
@@ -62,5 +66,27 @@ public async Task<GetAssetListResponse> GetListOfAssets(HousingSearchRequest que
...
+            var searchRequest = new HousingSearchRequest
+            {
+                SearchText = request.SearchText,
+                Page = request.Page,
+                PageSize = request.PageSize
+            };
+
+            var searchResponse = await _elasticSearchWrapper.Search<QueryableTransaction>(searchRequest).ConfigureAwait(false);
+
+            var loadedCount = searchResponse.Documents.Count;
```
Comment (amidomichael) [inline HousingSearchApi/V1/Interfaces/SearchGateway.cs:81]: can you combine `loadedCount` and `searchResponse.Documents.Select(queryableTransaction =>
                queryableTransaction.Create()` and pass them into the `GetTransactionListResponse` using `Create()` - or pass in `searchResponse` when creating the object and move all this logic into that class?

**Comment by amidomichael** [comment]
I would suggest making properties private where possible, in order to make the initialised object immutable. If you need to be able to change the value of a property, expose this via a public method and encapsulate the logic within the class itself.

**Review by amidomichael — APPROVED** [review]

File: HousingSearchApi/data/elasticsearch/transactionsIndex.json [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        "type": "keyword",
+        "normalizer": "my_normalizer",
+        "fields": {
+          "keyword": {
+            "type": "keyword",
+            "ignore_above": 256
+          }
+        }
+      },
+      "bankAccountNumber": {
```
Comment (LBHMGeorgieva) [inline HousingSearchApi/data/elasticsearch/transactionsIndex.json:61]: Is bank account number something that is needed for searching/displaying search results?

File: HousingSearchApi.Tests/V1/E2ETests/Steps/GetTenureSteps.cs [hunk 4]
```diff
@@ -41,16 +41,6 @@ public async Task WhenAPageSizeIsProvided(int pageSize)
             _lastResponse = await _httpClient.GetAsync(route).ConfigureAwait(false);
         }
 
-        public void ThenTheLastRequestShouldBeBadRequestResult()
```
Comment (LBHMGeorgieva) [inline HousingSearchApi.Tests/V1/E2ETests/Steps/GetTenureSteps.cs:44]: Hey Hanna, why were those removed (Same for assets)?

Comment (AnnaGolosova) [reply HousingSearchApi.Tests/V1/E2ETests/Steps/GetTenureSteps.cs:44]: I noticed that we have the same methods for all Steps. I moved this method to the `BaseSteps` class

File: HousingSearchApi.Tests/V1/E2ETests/Steps/ResponseModels/TransactionDTO.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+        public short FinancialMonth { get; set; }
+
+        public string TransactionSource { get; set; }
+
+        public Hackney.Shared.HousingSearch.Domain.Transactions.TransactionType TransactionType { get; set; }
+
+        public DateTime TransactionDate { get; set; }
+
+        public PersonDTO Person { get; set; }
```
Comment (LBHMGeorgieva) [inline HousingSearchApi.Tests/V1/E2ETests/Steps/ResponseModels/TransactionDTO.cs:25]: Hey Hanna, can you please confirm with me what the use case behind the person data for this endpoint is?

File: HousingSearchApi/V1/Boundary/Responses/Transactions/Person.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+{
+    public class Person
+    {
+        public Guid Id { get; }
+
+        public string FullName { get; }
+
+        private Person(Guid id, string fullName)
+        {
+            Id = id;
```
Comment (LBHMGeorgieva) [inline HousingSearchApi/V1/Boundary/Responses/Transactions/Person.cs:13]: Based on what Phillip was explaining, will we have a person ID for each transaction made? I think it might be best if we rename this to "Sender".

File: HousingSearchApi/V1/Interfaces/QueryGenerators/TransactionsQueryGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+        public QueryContainer Create(HousingSearchRequest request, QueryContainerDescriptor<QueryableTransaction> q)
+        {
+            if (string.IsNullOrWhiteSpace(request.SearchText)) return null;
+
+            return _queryBuilder
+                .WithWildstarQuery(request.SearchText,
+                    new List<string>
+                    {
+                        "person.fullName",
```
Comment (LBHMGeorgieva) [inline HousingSearchApi/V1/Interfaces/QueryGenerators/TransactionsQueryGenerator.cs:26]: Should that be sender?

File: HousingSearchApi/V1/Boundary/Responses/Transactions/TransactionResponse.cs [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        /// <summary>
+        /// Sender, who paid for the transaction
+        /// </summary>
+        /// <example>
+        /// {
+        ///     "Id": "6d290de9-75aa-46a9-8bf5-cb8e9bdf4ff0",
+        ///     "FullName": "Kian Hayward"
+        /// }
+        /// </example>
+        public Person Sender { get; }
```
Comment (LBHMGeorgieva) [inline HousingSearchApi/V1/Boundary/Responses/Transactions/TransactionResponse.cs:172]: I believe this property should no longer be of type "person"

**Review by LBHMGeorgieva — APPROVED** [review]
